Specialize optionals for grad_sum_to_size

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43743 Update aliasing in tensorexpr fuser
* #43742 use types in the IR instead of vmap
* #43636 Add inline call to profiling executor pipeline
* #43635 Use prim::TensorExprGroup interned symbol
* #43634 Update requires grad property
* **#43633 Specialize optionals for grad_sum_to_size**
* #43632 Add undefined specializations in backward
* #43631 refactor tests
* #43630 Refactor pass to class
* #43629 Add API for unexecuted op

In the backward graph, _grad_sum_to_size is inserted whenever a possibly broadcasting op is called:"
`"aten::_grad_sum_to_size(Tensor(a) self, int[]? size) -> Tensor(a)"`
 If a broadcast occurred, a sum is called, otherwise the second input is None and it is a no-op. Most of the time, it's a no-op (in the fast RNNs benchmark > 90% of the time). 

We can get rid of this op by profiling the optionality of the second input. I added `prim::profile_optional` to do this, which counts the number of times it saw a None value and the number of times it saw a value present. When specializing the backward graph, we insert checks for values we profiled as None, and in the optimized block can remove the grad_sum_to_size calls that use those values.

In the future we may revisit this when NNC supports reductions and we want to replace grad_sum_to_size with sums as well, but I think this is worth landing now.

Differential Revision: [D23358809](https://our.internmc.facebook.com/intern/diff/D23358809)

Commit:
Specialize optionals for grad_sum_to_size

[ghstack-poisoned]

Commit:
Update on "Specialize optionals for grad_sum_to_size"




[ghstack-poisoned]

eellison commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -1342,6 +1342,28 @@ struct ProfileOp : public Node {
 };
 
+struct TORCH_API ProfileOptionalOp : public Node {
+  static constexpr Symbol Kind = ::c10::prim::profile_optional;
```
It's a little unfortunate that this is basically a copy-pasta of the prim::profile node... I tried inheriting from it but ran into sum difficulties with the class Kind and some other APIs. Maybe in a follow up we can try to fix the inheritance and make the final prim::profile op that doesnt have any inputs its own op symbol. I think this is fine for now.

ZolotukhinM reviewed (approved):
Looks good to me, but maybe @Krovatkin should take another look at this.

Krovatkin commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -214,6 +233,31 @@ void ProfilingRecord::instrumentBlock(Block* block) {
+        opt_pn->setCallback(optional_profiler);
+        auto pno = opt_pn->addOutput();
+        pno->setType(i->type());
+        WithInsertPoint guard(i->node());
```
nitpick do we really need a `WithInsertPoint guard` here since we use `insertAfter`?

Krovatkin commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -214,6 +233,31 @@ void ProfilingRecord::instrumentBlock(Block* block) {
         insertShapeProfile(n, offset);
       }
+
+      if (i->type()->cast<OptionalType>() && hasGradSumToSizeUses(i)) {
```
nitpick: maybe a quick comment that we are profiling a definition vs uses `prim::profile`s profile?

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Specialize optionals for grad_sum_to_size"


In the backward graph, _grad_sum_to_size is inserted whenever a possibly broadcasting op is called:"
`"aten::_grad_sum_to_size(Tensor(a) self, int[]? size) -> Tensor(a)"`
 If a broadcast occurred, a sum is called, otherwise the second input is None and it is a no-op. Most of the time, it's a no-op (in the fast RNNs benchmark > 90% of the time). 

We can get rid of this op by profiling the optionality of the second input. I added `prim::profile_optional` to do this, which counts the number of times it saw a None value and the number of times it saw a value present. When specializing the backward graph, we insert checks for values we profiled as None, and in the optimized block can remove the grad_sum_to_size calls that use those values.

In the future we may revisit this when NNC supports reductions and we want to replace grad_sum_to_size with sums as well, but I think this is worth landing now.

Differential Revision: [D23358809](https://our.internmc.facebook.com/intern/diff/D23358809)

[ghstack-poisoned]

Commit:
Update on "Specialize optionals for grad_sum_to_size"


In the backward graph, _grad_sum_to_size is inserted whenever a possibly broadcasting op is called:"
`"aten::_grad_sum_to_size(Tensor(a) self, int[]? size) -> Tensor(a)"`
 If a broadcast occurred, a sum is called, otherwise the second input is None and it is a no-op. Most of the time, it's a no-op (in the fast RNNs benchmark > 90% of the time). 

We can get rid of this op by profiling the optionality of the second input. I added `prim::profile_optional` to do this, which counts the number of times it saw a None value and the number of times it saw a value present. When specializing the backward graph, we insert checks for values we profiled as None, and in the optimized block can remove the grad_sum_to_size calls that use those values.

In the future we may revisit this when NNC supports reductions and we want to replace grad_sum_to_size with sums as well, but I think this is worth landing now.

Differential Revision: [D23358809](https://our.internmc.facebook.com/intern/diff/D23358809)

[ghstack-poisoned]

eellison commented:
If anyone is investigating cuda thing, i have pr out here to fix tests, plz do not revert: https://github.com/pytorch/pytorch/pull/43633
